File: sdks/swift/Sources/NucleusSwift/Auth/NucleusAuth.swift
import Foundation
import Combine

/// Observable authentication state for the current user.
///
/// Use this object as an `@EnvironmentObject` or `@StateObject` in your SwiftUI
/// views to reactively update the UI when the user signs in or out.
///
/// ```swift
/// @StateObject private var auth = NucleusAuth()
/// ```
@MainActor
public final class NucleusAuth: ObservableObject {

    // MARK: - Published State

    /// The currently signed-in user, or `nil` if not authenticated.
    @Published public private(set) var user: NucleusUser?

    /// The active session, or `nil` if not authenticated.
    @Published public private(set) var session: NucleusSession?

    /// The user's active organization, or `nil`.
    @Published public private(set) var organization: NucleusOrganization?

    /// Whether the user is currently signed in with a valid session.
    public var isSignedIn: Bool {
        session != nil && user != nil
    }

    // MARK: - Private

    private let sessionManager = SessionManager.shared
    private let apiClient = APIClient.shared
    private var refreshTask: Task<Void, Never>?

    // MARK: - Init

    public init() {
        Task { await restoreSession() }
    }

    // MARK: - Public API

    /// Sign in with email and password.
    public func signIn(email: String, password: String) async throws {
        let response = try await SignInManager.shared.signIn(email: email, password: password)
        apply(response)
    }

    /// Start an OAuth flow for the given provider (e.g. `"google"`, `"github"`).
    public func signInWithOAuth(provider: String) async throws {
        let response = try await OAuthManager.shared.startOAuthFlow(provider: provider)
        apply(response)
    }

    /// Sign out the current user and clear stored tokens.
    public func signOut() async {
        do {
            if let token = session?.accessToken {
                try await apiClient.post(path: "/auth/sign-out", body: nil, token: token)
            }
        } catch {
            // Best-effort server sign-out; always clear local state.
        }

        sessionManager.clearSession()
        user = nil
        session = nil
        organization = nil
    }

    /// Returns the current access token, refreshing it if expired.
    public func getToken() async throws -> String {
        guard let currentSession = session else {
            throw NucleusAuthError.notAuthenticated
        }

        if currentSession.isExpired {
            try await refreshSession()
        }

        guard let token = session?.accessToken else {
            throw NucleusAuthError.notAuthenticated
        }

        return token
    }

    /// Switch to a different organization.
    public func switchOrganization(_ org: NucleusOrganization) async throws {
        let token = try await getToken()
        let body = ["organizationId": org.id]
        let data = try await apiClient.post(
            path: "/auth/switch-organization",
            body: body,
            token: token
        )
        let response = try JSONDecoder.nucleus.decode(AuthResponse.self, from: data)
        apply(response)
    }

    // MARK: - Internal

    private func restoreSession() async {
        guard let stored = sessionManager.loadSession() else { return }

        session = stored.session
        user = stored.user
        organization = stored.organization

        // Proactively refresh if the token is expired or about to expire.
        if stored.session.isExpired || stored.session.isExpiringSoon {
            try? await refreshSession()
        }
    }

    private func refreshSession() async throws {
        guard let refreshToken = session?.refreshToken else {
            throw NucleusAuthError.notAuthenticated
        }

        let response = try await sessionManager.refresh(refreshToken: refreshToken)
        apply(response)
    }

    private func apply(_ response: AuthResponse) {
        self.user = response.user
        self.session = response.session
        self.organization = response.organization
        sessionManager.saveSession(response)
    }
}

// MARK: - Errors

public enum NucleusAuthError: LocalizedError {
    case notAuthenticated
    case invalidCredentials
    case oauthFailed(String)
    case networkError(Error)
    case unknown(String)

    public var errorDescription: String? {
        switch self {
        case .notAuthenticated:
            return "User is not authenticated."
        case .invalidCredentials:
            return "Invalid email or password."
        case .oauthFailed(let reason):
            return "OAuth authentication failed: \(reason)"
        case .networkError(let error):
            return "Network error: \(error.localizedDescription)"
        case .unknown(let message):
            return message
        }
    }
}

// MARK: - Auth Response

struct AuthResponse: Codable {
    let user: NucleusUser
    let session: NucleusSession
    let organization: NucleusOrganization?
}

File: sdks/android/src/main/kotlin/dev/nucleus/auth/NucleusAuth.kt
package dev.nucleus.auth

import dev.nucleus.models.NucleusOrganization
import dev.nucleus.models.NucleusSession
import dev.nucleus.models.NucleusUser
import dev.nucleus.network.ApiClient
import dev.nucleus.session.SessionManager
import dev.nucleus.session.TokenStorage
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

/**
 * Central authentication state holder for the Nucleus SDK.
 *
 * Exposes [StateFlow] properties that Compose UIs can collect to
 * reactively respond to authentication changes.
 */
class NucleusAuth internal constructor(
    private val apiClient: ApiClient,
    private val tokenStorage: TokenStorage,
    private val sessionManager: SessionManager,
) {
    // ---- Reactive state (delegates to SessionManager) -------------------

    /** The currently authenticated user, or `null` if signed out. */
    val user: StateFlow<NucleusUser?> get() = sessionManager.user

    /** The active session, or `null` if signed out. */
    val session: StateFlow<NucleusSession?> get() = sessionManager.session

    /** Whether a user is currently signed in with a valid session. */
    val isSignedIn: StateFlow<Boolean> get() = sessionManager.isSignedIn

    // ---- Organization state ---------------------------------------------

    private val _organizations = MutableStateFlow<List<NucleusOrganization>>(emptyList())
    val organizations: StateFlow<List<NucleusOrganization>> = _organizations.asStateFlow()

    private val _activeOrganization = MutableStateFlow<NucleusOrganization?>(null)
    val activeOrganization: StateFlow<NucleusOrganization?> = _activeOrganization.asStateFlow()

    // ---- Actions --------------------------------------------------------

    /**
     * Sign in with email and password.
     *
     * @return the authenticated [NucleusUser] on success.
     * @throws NucleusAuthException on failure.
     */
    suspend fun signInWithEmailPassword(email: String, password: String): NucleusUser {
        val result = apiClient.signIn(email, password)
            ?: throw NucleusAuthException("Sign-in failed. Check credentials and try again.")
        sessionManager.setSession(result.session, result.user)
        return result.user
    }

    /** Sign out the current user and clear all local state. */
    suspend fun signOut() {
        try {
            apiClient.signOut()
        } catch (_: Exception) {
            // Best-effort server-side sign-out.
        }
        sessionManager.clearSession()
        _organizations.value = emptyList()
        _activeOrganization.value = null
    }

    /** Fetch the organizations that the current user belongs to. */
    suspend fun loadOrganizations(): List<NucleusOrganization> {
        val orgs = apiClient.getOrganizations()
        _organizations.value = orgs
        if (_activeOrganization.value == null && orgs.isNotEmpty()) {
            _activeOrganization.value = orgs.first()
        }
        return orgs
    }

    /** Switch the active organization context. */
    fun setActiveOrganization(organization: NucleusOrganization) {
        _activeOrganization.value = organization
    }

    // ---- Internal helpers -----------------------------------------------

    internal fun setSessionFromOAuth(session: NucleusSession, user: NucleusUser) {
        sessionManager.setSession(session, user)
    }
}

class NucleusAuthException(message: String, cause: Throwable? = null) :
    Exception(message, cause)
